### PR TITLE
Rewrite  *sh reduce unit tests to gtest: part 1

### PR DESCRIPTION
**Pull request by mozga-amd**
Currently tests (*._with_index) are written in .sh. This pull-request rewrites  *sh reduce unit tests to gtest.

### CODE REVIEW

File: test/reduce/reduce_with_index.cpp [hunk 1]
```diff
@@ -1,248 +1,163 @@
...
-            throw std::runtime_error("Invalid cmd-line arguments, more argumetns are needed!");
-
-        data_type   = std::atoi(argv[optind++]);
-        init_method = std::atoi(argv[optind]);
-
-        if(scales.empty())
+            _sett0.reduceOpId = ReduceTensorOp::AMAX;
+            _sett0.reduceDims = settReduceDim[i];
+            params.push_back(_sett0);
+            _sett0.reduceOpId = ReduceTensorOp::MIN;
```
Comment (aosewski) [inline test/reduce/reduce_with_index.cpp]: This is not used. Please add also reduction with other than `AMAX` op ids.

File: test/reduce/reduce_with_index.cpp [hunk 2]
```diff
@@ -1,248 +1,163 @@
...
-        int ch;
+    static std::vector<ReduceParam> params;
 
-        while(1)
+    static void SetUpTestSuite()
+    {
+        // set testcase variables
+        ReduceParam _sett0 = sett0();
+        // + reduce dims: Generic;
+        // set testcase variables
```
Comment (aosewski) [inline test/reduce/reduce_with_index.cpp]: Leftovers ?

File: test/reduce/reduce_with_index.cpp [hunk 3]
```diff
@@ -1,248 +1,163 @@
...
-                     "(only 1 or 3 or 4 dimensions supported)"
-                  << std::endl;
-        std::cout << "--scales or -S, comma separated two float values for alpha and beta"
-                  << std::endl;
-        std::cout << "Arg1 -- data type (1: fp32, 3: int8, 5: bp16, 6: fp64)" << std::endl;
-        std::cout << "Arg2 -- init method(0=no init, 1=single integer value, 2=scope integer "
-                     "value, 3=decimal value)"
-                  << std::endl;
-    };
+ReduceParam sett0()
```
Comment (aosewski) [inline test/reduce/reduce_with_index.cpp]: Maybe rename like this?
```suggestion
ReduceParam setDefault()
```

File: test/reduce/reduce_with_index.cpp [hunk 4]
```diff
@@ -1,248 +1,163 @@
...
-constexpr ReduceTensorOp reduceOpId = ReduceTensorOp::AMAX;
-constexpr bool propagateNan         = false;
+TYPED_TEST_SUITE(ReduceWithIndexFloat, Reduce_float_types);
+TYPED_TEST_SUITE(ReduceWithIndexDouble, Reduce_double_types);
+TYPED_TEST_SUITE(ReduceWithIndexInt8, Reduce_int8t_types);
+TYPED_TEST_SUITE(ReduceWithIndexHalf, Reduce_half_types);
+TYPED_TEST_SUITE(ReduceWithIndexBHalfFloat, Reduce_bhalf_float_Types);
 
-int main(int argc, char* argv[])
+TYPED_TEST(ReduceWithIndexFloat, ReduceWithIndexTestFloat)
```
Comment (aosewski) [inline test/reduce/reduce_with_index.cpp]: Maybe we could have different test cases for different reduce op ids ?

**Review by aosewski — CHANGES_REQUESTED** [review]

**Comment by aosewski** [comment]
@qianfengz Please take a look. Is that OK to change the reduction test script to unit-tests?

File: test/reduce/reduce_with_index.cpp (hunk 3 above)
Comment (bartekxk) [reply test/reduce/reduce_with_index.cpp]: You can just remove this function and add default values in the structure.

File: test/reduce/reduce_with_index.cpp [hunk 5]
```diff
@@ -1,248 +1,163 @@
...
-        if(optind + 2 > argc)
-            throw std::runtime_error("Invalid cmd-line arguments, more argumetns are needed!");
-
-        data_type   = std::atoi(argv[optind++]);
-        init_method = std::atoi(argv[optind]);
-
-        if(scales.empty())
+            _sett0.reduceOpId = ReduceTensorOp::AMAX;
+            _sett0.reduceDims = settReduceDim[i];
+            params.push_back(_sett0);
```
Comment (bartekxk) [inline test/reduce/reduce_with_index.cpp]: Remove set0 function and use emplace_back here

File: test/reduce/reduce_with_index.cpp [hunk 6]
```diff
@@ -1,248 +1,163 @@
...
+                param.do_dumpout,
+                param.time_kernel,
+                param.inLengths,
+                param.reduceDims,
+                param.reduceOpId,
+                param.propagateNan,
+                param.useIndex,
+                param.alpha,
+                param.beta);
+            // EXPECT_TRUE
```
Comment (bartekxk) [inline test/reduce/reduce_with_index.cpp]:
```suggestion
```

File: test/reduce/reduce_with_index.cpp (hunk 3 above)
Comment (mozga-amd) [reply test/reduce/reduce_with_index.cpp]: I agree with it. However, it is a conscious exclusion. In the next steps the following structure will be moved to a reduce file to resolve implicit calls.

Comment (bartekxk) [reply test/reduce/reduce_with_index.cpp]: So why not move it now? Which "reduce file" you mean?

Comment (mozga-amd) [reply test/reduce/reduce_with_index.cpp]: The pull-request contains only changes in this test: one case, one PR. Changes in structure interferes in API.

Comment (bartekxk) [reply test/reduce/reduce_with_index.cpp]: Sorry but I dont follow - why structure for test cases parameter require API changes? Maybe we can sync offline

File: test/reduce/reduce_no_index.cpp [hunk 7]
```diff
@@ -1,248 +1,148 @@
...
+    int init_method{2};
+    float alpha{1.0f};
+    float beta{0.0f};
+    std::vector<size_t> inLengths{64, 4, 280, 82};
+    std::vector<int> reduceDims{0, 1, 2, 3};
+    ReduceTensorOp reduceOpId{ReduceTensorOp::AMAX};
+};
 
-class SimpleAppArgs
+std::vector<std::vector<int>> settGenericReduceDim()
```
Comment (bartekxk) [inline test/reduce/reduce_no_index.cpp]:
```suggestion
std::vector<std::vector<int>> SetGenericReduceDim()
```

File: test/reduce/reduce_no_index.cpp [hunk 8]
```diff
@@ -1,248 +1,148 @@
...
-                     "(only 1 or 3 or 4 dimensions supported)"
-                  << std::endl;
-        std::cout << "--scales or -S, comma separated two float values for alpha and beta"
-                  << std::endl;
-        std::cout << "Arg1 -- data type (0: fp16, 1: fp32, 3: int8, 5: bp16, 6: fp64)" << std::endl;
-        std::cout << "Arg2 -- init method(0=no init, 1=single integer value, 2=scope integer "
-                     "value, 3=decimal value)"
-                  << std::endl;
+        // set testcase variables
+        ReduceParam _sett0;
```
Comment (bartekxk) [inline test/reduce/reduce_no_index.cpp]: Why you added underscore before variable name?

File: test/reduce/reduce_with_index.cpp [hunk 9]
```diff
@@ -1,248 +1,148 @@
...
+    int init_method{2};
+    float alpha{1.0f};
+    float beta{0.0f};
+    std::vector<size_t> inLengths{64, 4, 280, 82};
+    std::vector<int> reduceDims{0, 1, 2, 3};
+    ReduceTensorOp reduceOpId{ReduceTensorOp::AMAX};
+};
 
-class SimpleAppArgs
+std::vector<std::vector<int>> settGenericReduceDim()
```
Comment (bartekxk) [inline test/reduce/reduce_with_index.cpp]:
```suggestion
std::vector<std::vector<int>> SetGenericReduceDim()
```

File: test/reduce/reduce_with_index.cpp [hunk 10]
```diff
@@ -1,248 +1,148 @@
...
-                     "(only 1 or 3 or 4 dimensions supported)"
-                  << std::endl;
-        std::cout << "--scales or -S, comma separated two float values for alpha and beta"
-                  << std::endl;
-        std::cout << "Arg1 -- data type (1: fp32, 3: int8, 5: bp16, 6: fp64)" << std::endl;
-        std::cout << "Arg2 -- init method(0=no init, 1=single integer value, 2=scope integer "
-                     "value, 3=decimal value)"
-                  << std::endl;
+        // set testcase variables
+        ReduceParam _sett0;
```
Comment (bartekxk) [inline test/reduce/reduce_with_index.cpp]: Why you added underscore before variable name?

**Review by bartekxk — COMMENTED** [review]
LGTM, minor comments

**Review by bartekxk — DISMISSED** [review]
Please merge after CI pass

File: test/reduce/reduce_no_index.cpp [hunk 11]
```diff
@@ -1,248 +1,148 @@
...
-constexpr ReduceTensorOp reduceOpId = ReduceTensorOp::AVG;
-constexpr bool propagateNan         = false;
+TYPED_TEST_SUITE(ReduceWithIndexFloat, Reduce_float_types);
+TYPED_TEST_SUITE(ReduceWithIndexDouble, Reduce_double_types);
+TYPED_TEST_SUITE(ReduceWithIndexInt8, Reduce_int8t_types);
+TYPED_TEST_SUITE(ReduceWithIndexHalf, Reduce_half_types);
+TYPED_TEST_SUITE(ReduceWithIndexBHalfFloat, Reduce_bhalf_float_Types);
 
-int main(int argc, char* argv[])
+TYPED_TEST(ReduceWithIndexFloat, ReduceWithIndexTestFloat)
```
Comment (aosewski) [inline test/reduce/reduce_no_index.cpp]: Can we have here separate test cases for each reduce op? If any of them will fail we would get clear result indicating what exactly failed.

File: test/reduce/reduce_no_index.cpp [hunk 12]
```diff
@@ -1,248 +1,148 @@
...
+
+        for(std::size_t i(0); i < setReduceDim.size(); ++i)
+        {
+            set.reduceDims = setReduceDim[i];
+            set.reduceOpId = ReduceTensorOp::AMAX;
+            params.emplace_back(set);
+            set.reduceOpId = ReduceTensorOp::MIN;
+            params.emplace_back(set);
+            set.reduceOpId = ReduceTensorOp::MAX;
+            params.emplace_back(set);
```
Comment (aosewski) [inline test/reduce/reduce_no_index.cpp]: I'd prefer if we have `reduceOpId` as a parameter to `Run` function. See comment below.

File: test/reduce/reduce_with_index.cpp [hunk 13]
```diff
@@ -56,17 +41,18 @@ class ReduceWithIndexTest : public ::testing::Test
...
-            _sett0.reduceDims = settReduceDim[i];
-            params.push_back(_sett0);
-            _sett0.reduceOpId = ReduceTensorOp::MIN;
+            set.reduceDims = setReduceDim[i];
+            set.reduceOpId = ReduceTensorOp::AMAX;
+            params.emplace_back(set);
+            set.reduceOpId = ReduceTensorOp::MIN;
+            params.emplace_back(set);
+            set.reduceOpId = ReduceTensorOp::MAX;
+            params.emplace_back(set);
```
Comment (aosewski) [inline test/reduce/reduce_with_index.cpp]: As above.

**Review by aosewski — CHANGES_REQUESTED** [review]
I'd split each test case into separate `ReduceOpIds`. Other than that looks good.

**Review by aosewski — DISMISSED** [review]
Approving just in case my request would block merge. @bartekxk please review after adding mentioned changes.

**Review by junliume — DISMISSED** [review]

File: test/reduce/reduce_no_index.cpp [hunk 14]
```diff
@@ -1,248 +1,205 @@
+
+
 // SPDX-License-Identifier: MIT
```
Comment (bartekxk) [inline test/reduce/reduce_no_index.cpp]:
```suggestion
// SPDX-License-Identifier: MIT
```

**Review by bartekxk — DISMISSED** [review]

**Review by bartekxk — APPROVED** [review]

**Pull request merged**